pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -0,0 +1,183 @@
+}
+
+//HintProvider interface is to be implemented by Hint Providers
+type HintProvider interface {
+	GetTopologyHints(pod v1.Pod, container v1.Container) ([]TopologyHint, bool)
+}
+
+//Store interface is to allow Hint Providers to retrieve pod affinity
+type Store interface {
+	GetAffinity(podUID string, containerName string) TopologyHint
+}
+
+type containers map[string]TopologyHint
+
+var _ Manager = &manager{}
+
+type policyName string
+
+//NewManager creates a new TopologyManager based on provided policy
+func NewManager(topologyPolicyName string) Manager {
+	klog.Infof("[topologymanager] Creating topology manager with %s policy", topologyPolicyName)
+	var policy Policy
+
+	switch policyName(topologyPolicyName) {
+
+	case PolicyPreferred:
+		policy = NewPreferredPolicy()
+
+	case PolicyStrict:
+		policy = NewStrictPolicy()
+
+	default:
+		klog.Errorf("[topologymanager] Unknow policy %s, using default policy %s", topologyPolicyName, PolicyPreferred)
+		policy = NewPreferredPolicy()
+	}
+
+	var hp []HintProvider
+	pnh := make(map[string]containers)
+	pm := make(map[string]string)
+	manager := &manager{
+		hintProviders:    hp,
+		podTopologyHints: pnh,
+		podMap:           pm,
+		policy:           policy,
+	}
+
+	return manager
+}
+
+func (m *manager) GetAffinity(podUID string, containerName string) TopologyHint {
+	return m.podTopologyHints[podUID][containerName]
+}
+
+func (m *manager) calculateTopologyAffinity(pod v1.Pod, container v1.Container) (TopologyHint, bool) {
+	socketMask := socketmask.NewSocketMask(nil)
+	var maskHolder []string
+	var socketMaskInt64 [][]int64
+	count := 0
+	admitPod := true
+	for _, hp := range m.hintProviders {
+		topologyHints, admit := hp.GetTopologyHints(pod, container)
+		for r := range topologyHints {
+                	socketMaskVals := []int64(topologyHints[r].SocketMask)
+                        socketMaskInt64 = append(socketMaskInt64,socketMaskVals)
+                }
+		if !admit && topologyHints == nil {
+            		klog.Infof("[topologymanager] Hint Provider does not care about this container")
+            		continue
+        	}
+		if admit && topologyHints != nil {
+			socketMask, maskHolder = socketMask.GetSocketMask(socketMaskInt64, maskHolder, count)
+			count++
+		} else if !admit && topologyHints != nil {
+			klog.Infof("[topologymanager] Cross Socket Topology Affinity")
+			admitPod = false
+			socketMask, maskHolder = socketMask.GetSocketMask(socketMaskInt64, maskHolder, count)
+			count++
+		}
+	}
+	var topologyHint TopologyHint
+	topologyHint.SocketMask = socketMask 
+
+	return topologyHint, admitPod
+}
+
+func (m *manager) AddHintProvider(h HintProvider) {
+	m.hintProviders = append(m.hintProviders, h)
+}
+
+func (m *manager) AddContainer(pod *v1.Pod, containerID string) error {
+	m.podMap[containerID] = string(pod.UID)
+	return nil
+}
+
+func (m *manager) RemoveContainer(containerID string) error {
+	podUIDString := m.podMap[containerID]
+	delete(m.podTopologyHints, podUIDString)
```

Comment:
Hmm, this won't quite work because `m.podTopologyHints[podUIDString]` isn't indexed by `containerID`, but rather by `containerName`. Not sure immediately how to address this...

Reply:
Going to postpone a solution for this for now